transport/x/cherami/outbound.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/serialize"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+var errOutboundNotStarted = errors.ErrOutboundNotStarted("cherami.Outbound")
+
+// OutboundConfig defines the config in order to create a Outbound.
+// if frontend is provided, we'll connect to the provided frontend
+// otherwise, hyperbahn will be used to connect to cherami
+type OutboundConfig struct {
+	Destination string
+	Frontend    string
+	Port        int
+}
+
+// Outbound is a outbound that uses cherami as the transport
+type Outbound struct {
+	config         OutboundConfig
+	publisher      cherami.Publisher
+	tracer         opentracing.Tracer
+	cheramiFactory CheramiFactory
+
+	once sync.LifecycleOnce
+}
+
+type receipt struct{ cherami.PublisherReceipt }
+
+func (r receipt) String() string {
+	return r.Receipt
+}
+
+// NewOutbound builds a new cherami outbound
+func NewOutbound(config OutboundConfig) *Outbound {
```

Comment:
This should be a function on a "Cherami.Transport" similar to how http.NewOutbound and tchannel.NewOutbound function.